RecoCTPPS/PixelLocal/src/RPixPlaneCombinatoryTracking.cc

```diff
@@ -0,0 +1,513 @@
+
+RPixPlaneCombinatoryTracking::RPixPlaneCombinatoryTracking(edm::ParameterSet const& parameterSet) : 
+  RPixDetTrackFinder(parameterSet){
+
+  trackMinNumberOfPoints_ = parameterSet.getParameter<uint>("trackMinNumberOfPoints");
+  verbosity_ = parameterSet.getUntrackedParameter<int> ("verbosity");
+  maximumChi2OverNDF_ = parameterSet.getParameter<double>("maximumChi2OverNDF");
+  maximumXLocalDistanceFromTrack_= parameterSet.getParameter<double>("maximumXLocalDistanceFromTrack");
+  maximumYLocalDistanceFromTrack_= parameterSet.getParameter<double>("maximumYLocalDistanceFromTrack");
+  numberOfPlanesPerPot_ = parameterSet.getParameter<int> ("numberOfPlanesPerPot"   );
+
+  if(trackMinNumberOfPoints_<3){
+    throw cms::Exception("RPixPlaneCombinatoryTracking") << "Minimum number of planes required for tracking is 3, "
+                                                         << "tracking is not possible with " 
+                                                         << trackMinNumberOfPoints_ << " hits";
+  }
+ 
+  
+}
+
+//------------------------------------------------------------------------------------------------//
+    
+RPixPlaneCombinatoryTracking::~RPixPlaneCombinatoryTracking() {
+  possiblePlaneCombinations_.clear();
+}
+
+//------------------------------------------------------------------------------------------------//
+
+void RPixPlaneCombinatoryTracking::initialize(){
+ 
+  uint32_t numberOfCombinations = factorial(numberOfPlanesPerPot_)/
+                                  (factorial(numberOfPlanesPerPot_-trackMinNumberOfPoints_)
+                                   *factorial(trackMinNumberOfPoints_));
+  if(verbosity_>=2) edm::LogInfo("RPixPlaneCombinatoryTracking")<<"Number of combinations = "<<numberOfCombinations;
+  possiblePlaneCombinations_.reserve(numberOfCombinations);
+
+  possiblePlaneCombinations_ = getPlaneCombinations(listOfAllPlanes_,trackMinNumberOfPoints_);
+
+  if(verbosity_>=2) {
+    for( const auto & vec : possiblePlaneCombinations_){
+      for( const auto & num : vec){
+        edm::LogInfo("RPixPlaneCombinatoryTracking")<<num<<" - ";
+      }
+      edm::LogInfo("RPixPlaneCombinatoryTracking");
+    }
+  }
+
+}
+
+//------------------------------------------------------------------------------------------------//
+    
+//This function produces all the possible plane combinations extracting numberToExtract planes over numberOfPlanes planes
+RPixPlaneCombinatoryTracking::PlaneCombinations
+RPixPlaneCombinatoryTracking::getPlaneCombinations(const std::vector<uint32_t> &inputPlaneList, uint32_t numberToExtract)
+{
+    uint32_t numberOfPlanes = inputPlaneList.size();
+    std::string bitmask(numberToExtract, 1); // numberToExtract leading 1's
+    bitmask.resize(numberOfPlanes, 0); // numberOfPlanes-numberToExtract trailing 0's
+    PlaneCombinations planeCombinations;
+
+    // store the combination and permute bitmask
+    do {
+      std::vector<uint32_t> tmpPlaneList;
```

Comment:
please avoid this intermediate memory allocation. you can first push back onto planeCombinations and get a const& to keep the code simple